### PR TITLE
Data wn18

### PR DESCRIPTION
**Pull request by igorgarbuz**
PR intended for the very 1st review. Modifications are planned according to ongoing changes regarding composite arrows computation. Any kind of feedback is welcome.

### CODE REVIEW

File: catlearn/data/utils.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+        elif isinstance(file, set):
+            for v in file:
+                f.write(str(v) + '\n')
+        else:
+            raise ValueError('Unsupported datatype for writing to file.')
+
+def get_entities_relations(ds_list: List[str]) -> [List[str], List[str]]:
+    """Get all entities and relations as 2 separate lists. Used to generate sets of unique values."""
+    triplets = [line.split(raw_dataset_pat['sep']) for line in ds_list]
+    return [h.strip() for h, r, t in triplets] + [t.strip() for h, r, t in triplets], [r.strip() for h, r, t in triplets]
```
Comment (Previsou) [inline catlearn/data/utils.py:60]: Could be done as `tuple(zip(*triplets))`, except for the stripping part though - I guess that one can also be treated easily in a previous line.

Whatever the selected code, I tend to favor using `_` for unused variables in list comprehension

File: catlearn/data/utils.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+    Args:
+        path_word2vec (str): path to word2vec model file location.
+        Either .txt or .pkl
+        path_entity_dic (str): path to the directory with entity2id file location
+        File name 
+    """
+    
+    word_set: List[str] = read_file(os.path.join(path_dir_wordset, preproc_pat['wordset']))
+    import word2vec
```
Comment (Previsou) [inline catlearn/data/utils.py:129]: Never import within a function

File: catlearn/data/utils.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    import torch
+    w2v_model = word2vec.load(path_word2vec)
+
+    w2v_short: Dict = {}
+    vec_dim = len(w2v_model['the'])
+    for k in word_set:
+        try: w2v_short[k] = w2v_model[k]
+        except:
```
Comment (Previsou) [inline catlearn/data/utils.py:137]: Never used raw `expect:` as it also catches synrax errors! At least go for `except Exception`

File: catlearn/data/utils.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+            try: w2v_short[k] = w2v_model[k.split('_')[-1]]
+            except:
+                try: w2v_short[k] = w2v_model[k.split['_'][-2]]
```
Comment (Previsou) [inline catlearn/data/utils.py:140]: hum should be `k.split('_')`

**Review by Previsou — CHANGES_REQUESTED** [review]
Haven't finished yet. I'll continue tomorrow :-)

File: scripts/train.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    @property
+    def parameters(self) -> Callable[[], Iterable[Any]]:
+        return self.linear.parameters
+
+    def __call__(self, x: Tsor, y: Tsor, l: Tsor) -> Tsor:
+        """ Compute x R y """
+        return self.linear(torch.cat((x, y, l), -1))
+
+class CustomScore(ScoringModel):
+    """ Must be defined. depends on labebra and
```
Comment (Previsou) [inline scripts/train.py:69]: Typo: labebra

File: scripts/train.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+            self,
+            nb_features: int,
+            nb_scores: int,
+            algebra: Algebra) -> None:
+        self.linear = torch.nn.Linear(
+            2 * nb_features + algebra.flatdim, nb_scores + 1)
+        self.softmax = torch.nn.Softmax(dim=-1)
+
+    @property
+    def parameters(self) -> Callable[[], Iterable[Any]]:
```
Comment (Previsou) [inline scripts/train.py:81]: Isn't this property already defined for ScoringModel?

File: scripts/train.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+        return self.linear.parameters
+
+    def __call__(self, src: Tsor, dst: Tsor, rel: Tsor) -> Tsor:
+        """ Compute S(src, dst, rel) """
+        cat_input = torch.cat((src, dst, rel), -1)
+        return self.softmax(self.linear(cat_input))[..., :-1]
+
+def train(ds_name: str, ds_path: str):
+    """[summary]
```
Comment (Previsou) [inline scripts/train.py:90]: Missing docstring?

File: scripts/train.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+    """
+    if ds_name == 'wn18':
+        ds = Dataset(path=ds_path, ds_name='wn18', node_vec_dim=10)
+    elif ds_name == 'fb15':
+        ds = Dataset(path=ds_path, ds_name='fb15', node_vec_dim=10)
```
Comment (Previsou) [inline scripts/train.py:95]: Should we raise an exception for unknwon datasets?

File: scripts/train.py [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+    train_step_log('W', f'Loaded dataste {ds_name}')
+    labels = DirectedGraph(ds.train)
+    train_step_log('W', f'Created DirectedGraph')
+    algebra=VectAlgebra(ds.entity_vec_dim)
```
Comment (Previsou) [inline scripts/train.py:99]: Minor: pylint would require you to add spaces around `=`

File: scripts/train.py [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+    parser.add_argument('--ds_path', help=('Path to a dataset folder where raw dataset files are stored.'
+                                        'For raw files naming format check config.py.'),
+                        type=str, default='')
+    return parser
+
+if __name__ == '__main__':
+    parser = parse_args()
+    try:
+        args = parser.parse_args()
+    except :
```
Comment (Previsou) [inline scripts/train.py:154]: Always specify an exception type, otherwise you also catch SyntaxErrors

File: catlearn/data/utils.py [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+
+    Args:
+        path (str): path where raw dataset files 'train.txt', 'valid.txt',
+        and 'test.txt' are stored
+    """
+    train: List[str] = read_file(os.path.join(path, raw_dataset_pat['train']))
+    entities, relations = get_entities_relations(train)
+    entities_set = set(entities)
+    relations_set = set(relations)
+    # Check validation set on entities and relations not present in the training set
```
Comment (Previsou) [inline catlearn/data/utils.py:77]: Minor: the valid and test set validation can be factored

File: catlearn/data/utils.py [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+            logging.warning(warning_str)
+            entities_set.add(e_test)
+    for r_test in set(relations_test):
+        if not r_test in relations_set:
+            warning_str = f'Relation ${r_test} present in testing set is not present in training set. It was added.'
+            warnings.warn(str_color('W', warning_str), UserWarning)
+            logging.warning(warning_str)
+            relations_set.add(r_test)
+    # Create dictionary with index
+    entity2id: Dict = {e: id for e, id in zip(entities_set, range(len(entities_set)))}
```
Comment (Previsou) [inline catlearn/data/utils.py:108]: Can simply be: ` entity2id = dict(enumerate(entities_set))`

File: catlearn/data/utils.py [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+    # id2entity = {id: e for e, id in entity2id.items()}
+    relation2id: Dict = {r: id for r, id in zip(relations_set, range(len(relations_set)))}
+    word2id: set = set([w.split('.')[0] for w in entities_set])
```
Comment (Previsou) [inline catlearn/data/utils.py:111]: The var implies a mapping similar to other entities, but we have only a set - and we write the set in the file

**Comment by Previsou** [comment]
In addition, you have lots of pylint warnings, cleanup. While the whole code is not 100% clean, we strive to have as little as possible such linting errors.

You can have a look with `pylint -f colorized (catlearn|scripts|tests)`

File: catlearn/data/utils.py [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+        and 'test.txt' are stored
+    """
+    train: List[str] = read_file(os.path.join(path, raw_dataset_pat['train']))
+    entities, relations = get_entities_relations(train)
+    entities_set = set(entities)
+    relations_set = set(relations)
+    # Check validation set on entities and relations not present in the training set
+    valid: List[str] = read_file(os.path.join(path, raw_dataset_pat['valid']))
+    entities_val, relations_val = get_entities_relations(valid)
+    for e_val in set(entities_val):
```
Comment (Kuraao) [inline catlearn/data/utils.py:80]: add some space around the loops, it is too packed for good readability

File: catlearn/data/utils.py (hunk 2 above)
Comment (Kuraao) [reply catlearn/data/utils.py:129]: unless there are very specific side effects from this import statement, move this to the beginning of the file
(order of imports: standard lib, external deps, other files of the project)

File: catlearn/data/utils.py (hunk 3 above)
Comment (Kuraao) [reply catlearn/data/utils.py:137]: As Maxime points out raw except catches syntax errors.
Also the expected exception is clear here (KeyError)

And this seems to me that this try, except is used for control flow, which is bad practice. use if/else statements.

File: catlearn/data/utils.py [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+    # id2entity = {id: e for e, id in entity2id.items()}
+    relation2id: Dict = {r: id for r, id in zip(relations_set, range(len(relations_set)))}
+    word2id: set = set([w.split('.')[0] for w in entities_set])
+    write_file(os.path.join(path, preproc_pat['wordset']), word2id)
+    # word2id: Dict = {k, id for k, id in zip(entities_set, range(len()))}
+    write_file(os.path.join(path, preproc_pat['relation2id']), relation2id)
+    write_file(os.path.join(path, preproc_pat['entity2id']), entity2id)
+
+
+def preprocess_word2vec(path_word2vec: str, path_dir_wordset: str):
```
Comment (Kuraao) [inline catlearn/data/utils.py:118]: Typing declaration for functions without returns:
def func_name(*args, **kwargs)-> None:

**Review by Kuraao — CHANGES_REQUESTED** [review]
additional comment:
the pathlib lib generally makes for cleaner code for path manipulations compared to os.path

File: catlearn/data/utils.py (hunk 12 above)
Comment (Previsou) [reply catlearn/data/utils.py]: Sorry about that, I realise now the solution I proposed would create the reverse of what you wanted. Can be `entity2id = {e: id for id, e in enumerate(entities_set)}`

**Comment by Previsou** [comment]
@igorgarbuz would you mind trying to clean up the branch? Seeing Chris's and my commits in your branch makes everything harder to review, and you would conflict anyway

**Comment by igorgarbuz** [comment]
Will squash it. 
> @igorgarbuz would you mind trying to clean up the branch? Seeing Chris's and my commits in your branch makes everything harder to review, and you would conflict anyway
